src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs

```diff
@@ -30,5 +36,71 @@ internal static class SemanticClassificationCacheUtilities
 
             return (documentKey, checksums.Text);
         }
+
+        public static async Task AddSemanticClassificationsAsync(
+            Document document,
+            TextSpan textSpan,
+            IClassificationService classificationService,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            bool isRazorDoc,
+            CancellationToken cancellationToken)
+        {
+            var workspaceStatusService = document.Project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+
+            // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
+            // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
+            // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
+            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
+            // fully loaded.
+            var isFullyLoadedTask = workspaceStatusService.IsFullyLoadedAsync(cancellationToken);
+            var isFullyLoaded = isFullyLoadedTask.IsCompleted && isFullyLoadedTask.GetAwaiter().GetResult();
+
+            // If we're not fully loaded try to read from the cache instead so that classifications appear up to date.
+            // New code will not be semantically classified, but will eventually when the project fully loads.
+            if (await TryAddSemanticClassificationsFromCacheAsync(document, textSpan, classifiedSpans, isFullyLoaded, cancellationToken).ConfigureAwait(false))
+                return;
+
+            // We need to special case Razor. Since the C# syntactic classifier doesn't run on their end, we need to
```

Comment:
This if-check and the added `isRazorDoc` parameter are the only new parts of these methods; they were otherwise moved untouched from `SemanticClassificationUtilities.cs`. As previously mentioned, the `isRazorDoc` parameter was added for testing purposes.